### ISSUE TITLE
About test results

### ISSUE THREAD

**Issue by huamo555** [ManipTrans/ManipTrans#18]
Hello Author:
Your outstanding work is fresh in my mind and I am now reproducing your work.
1, I have completed the training for the first task, where can I get the indicator after I perform the TEST operation?
2, For all the tasks, do I need to train and test them individually?
3, How can I get the final evaluation metrics in your paper?

![Image](https://github.com/user-attachments/assets/224ce84e-2550-4109-bd3a-d8ccb8581d04)

**Comment by KailinLi** [comment]
Hi there,

Thank you so much for your interest in ManipTrans, and sorry for the delayed response.

Because of the current test pipeline and the way the validation data are organized, we don’t yet have a “one-click” scoring script—sorry for the inconvenience! Below is a step-by-step recipe.

---

### 1.  Get the validation sequences  
We evaluate on a **filtered** OakInk-V2 *validation* split (oversize objects, very noisy clips, and some meaningless motions have been manually removed). After downloading the `DexManipNet` data, run

```bash
python DexManipNet/get_val_seq.py
```

This returns a list of dictionaries such as  
```python
{'dataIndices': '[99f9d@1]', 'side': 'BiH', 'rolloutLen': 623, 'rolloutBegin': 0}
```

---

### 2.  Train one policy per sequence
For every `dataIndices` in the validation list, train an individual residual policy and save its checkpoint.

---

### 3.  Patch the failure-detection logic
Please replace the original failure test in [single hand](https://github.com/ManipTrans/ManipTrans/blob/649d9ad53b708afdb1742786e5b4f08cdf3db027/maniptrans_envs/lib/envs/tasks/dexhandmanip_sh.py#L1528) and [bimanual](https://github.com/ManipTrans/ManipTrans/blob/649d9ad53b708afdb1742786e5b4f08cdf3db027/maniptrans_envs/lib/envs/tasks/dexhandmanip_bih.py#L1959) tasks with:

```python
       failed_execute = (
           (
               (diff_obj_pos_dist > 0.03)
               | (diff_thumb_tip_pos_dist > 0.06)
               | (diff_index_tip_pos_dist > 0.06)
               | (diff_middle_tip_pos_dist > 0.06)
               | (diff_pinky_tip_pos_dist > 0.06)
               | (diff_ring_tip_pos_dist > 0.06)
               | (diff_level_1_pos_dist > 0.08)
               | (diff_level_2_pos_dist > 0.08)
               | (diff_obj_rot_angle.abs() / np.pi * 180 > 30)
           )
           & (running_progress_buf >= 8)
       ) | error_buf
```

We’ll release a cleaner fix so manual patching won’t be necessary.

---

### 4.  Export the rollout states used for scoring

Run the command below after substituting `dataIndices`, `side`, `rolloutLen`, and `rolloutBegin` with the values obtained in *Step 1*. (See GitHub [issue #8](https://github.com/ManipTrans/ManipTrans/issues/8) for details on the exported data format.)

```bash
python main/rl/train.py task=ResDexHand dexhand=inspire side={side_placeholder} headless=true num_envs=256 test=true randomStateInit=false dataIndices={dataIndices_placeholder} rh_base_model_checkpoint=assets/imitator_rh_inspire.pth lh_base_model_checkpoint=assets/imitator_lh_inspire.pth checkpoint={checkpoint_placeholder} actionsMovingAverage=0.4 rolloutLen={rolloutLen_placeholder} rolloutBegin={rolloutBegin_placeholder} experiment={tag_placeholder_wfail} save_rollouts=true num_rollouts_to_save=128 num_rollouts_to_run=8192 save_successful_rollouts_only=false
```

---

### 5.  Calculate the success rate
Simply take the fraction of *successful* trajectories stored in the exported `hdf5` files.

---

### 6.  Compute the error metrics (`E_r`, `E_t`, `E_j`, `E_ft`)  
Compare the **successful** rollout state to the ground-truth pose from the MoCap dataset.
**Remember to align the time window**—because we sliced the rollout to `[rolloutBegin, rolloutBegin+rolloutLen)` in *Step 4*, apply the same offset to the GT sequence before evaluation. Pseudo-code:

```python
       def error_r(gt, rollout):
           # gt: [T, 3, 3]
           # rollout: [T, 4] (x, y, z, w)
           rollout = rollout[:, [3, 0, 1, 2]] # -> (w, x, y, z)
           rollout = quat_to_rotmat(rollout)
           res = gt @ rollout.transpose(-1, -2)
           res = rotmat_to_aa(res)
           diff_angle = torch.norm(res, dim=1)
           diff_angle = torch.min(diff_angle, 2 * np.pi - diff_angle)
           diff_angle = diff_angle / np.pi * 180
           return diff_angle.mean(0)
         
       def error_t(gt, rollout):
           # gt: [T, 3]
           # rollout: [T, 3]
           res = torch.norm(gt - rollout, dim=1)
           return res.mean(0)
   
       def error_j(gt, rollout):
           # gt: [T, F, 3]
           # rollout: [T, F, 3]
           res = torch.norm(gt - rollout, dim=2)
           return res.mean()
   
       def error_ft(gt, rollout):
           # gt: [T, F, 3]
           # rollout: [T, F, 3]
           fingertip_idx = [3, 6, 9, 12, 17] # inspire hand finger tips
           gt_tips = gt[:, fingertip_idx]
           rollout_tips = rollout[:, fingertip_idx]
           res = torch.norm(gt_tips - rollout_tips, dim=2)
           return res.mean()
```

---

### 7.  Aggregate over the validation set
Report the mean error per sequence, then average those across the entire split.
For bimanual data, first average the left & right hand errors within each sequence and then average over all $N$ sequences: $\frac{1}{N}\sum_n(\frac{\frac{1}{T}\sum_t(E_{\text{left}})+\frac{1}{T}\sum_t(E_{\text{right}})}{2})$

---

That’s it! We know the current workflow is still rather manual, and we’re actively working on a cleaner “one-click” scorer.

Feel free to ask if anything is unclear—happy to help!

**Comment by huamo555** [comment]
Hello author, thank you for your reply. I found that the **oakinkv2_val_list.json** file is missing when I run the **get_val_seq.py** file in step 1, can you tell me where this file is stored?

**Comment by KailinLi** [comment]
Hi, thanks for reaching out.

After you download **DexManipNet**, please unzip the package—you’ll find **oakinkv2_val_list.json** at  
`dexmanipnet/dexmanipnet_oakinkv2/oakinkv2_val_list.json`.  
You can also run our dataset-visualization script as a quick sanity check. Hope this helps!

**Comment by huamo555** [comment]
Hi author, I have now trained 3000 rounds on the [c9cd3@1] task, but after doing the test, I found that the captures saved in the dump are all **failed**, what could be the reason for this?

![Image](https://github.com/user-attachments/assets/fc7a7054-8dce-482b-97a2-bcbae76a92d2)

![Image](https://github.com/user-attachments/assets/ae7e8f39-3ded-4426-a6aa-f285372f7398)

The output reward at the time of the test was：

![Image](https://github.com/user-attachments/assets/89a2c029-0ff9-4ba6-8393-4b4c9e2a7b3c)

The display is as follows：

https://github.com/user-attachments/assets/141a6608-b92d-43a5-8ee0-e741abc3cd4d

**Comment by KailinLi** [comment]
Hi, it's difficult to pinpoint the exact issue from such a short snippet.

You can refer to this issue to check your training logic: [here](https://github.com/ManipTrans/ManipTrans/issues/22#issuecomment-2826796932).

Additionally, here are three more things to consider:
1. When evaluating, pass `rolloutLen` and `rolloutBegin` (from the first step) when saving rollout results.
2. You can set `save_successful_rollouts_only=true` to only save successful sequences and increase `num_rollouts_to_save` and `num_rollouts_to_run` to attempt more trials.
3. In some cases, certain sequences can be difficult to learn, and it may not be possible to transfer them. However, this should not occur too frequently.

**Comment by huamo555** [comment]
> 嗨，很难从这么短的片段中确定确切的问题。
> 
> 你可以参考这个问题来检查你的训练逻辑：[这里](https://github.com/ManipTrans/ManipTrans/issues/22#issuecomment-2826796932)。
> 
> 此外，还有三件事需要考虑：
> 
> 1. 评估时，在保存转出结果时传递 and（从第一步开始）。`rolloutLen``rolloutBegin`
> 2. 您可以设置为仅保存成功的序列，并增加并尝试更多试验。`save_successful_rollouts_only=true``num_rollouts_to_save``num_rollouts_to_run`
> 3. 在某些情况下，某些序列可能难以学习，并且可能无法传输它们。但是，这种情况不应过于频繁。

I tried training [e76b2@2] and [d7c33@1] again, and the training process for these two tasks is the same as the training setup for [c9cd3@1], and strangely, the model succeeds on basically all of them, so it may be that the sequence [c9cd3@1] is difficult to learn. I have two more questions at the moment:
1, roughly how many rounds do you train when training, am I training 3000 rounds too many.
2, Do you have any plans to write a code for one-click training and testing? This will make it easy to reproduce.

Thank you in advance for your response!

**Comment by KailinLi** [comment]
Hi there,

Thank you for taking the time to test our code and share thoughtful questions.

MoCap noise, object dynamics, and the morphological gap between a human hand and a dexterous hand can make some sequences difficult to transfer, so a more tailored approach is sometimes necessary.

Regarding your points:

1. In our experiments we typically run about 800 – 1,800 training iterations; 3,000 is usually more than a sequence needs.  
2. I’ve recently taken on several new projects, so my schedule is quite full. I’ll try to prepare a one-click testing script during my spare moments, though I can’t commit to a specific date. If you’re aiming for a result in the next 2-3 weeks, you might consider following the workflow we previously outlined and adapting it to your setup.

Thanks again for your interest and for helping improve the project.

**Comment by huamo555** [comment]
Thank you for your reply, I will close the issue.
